[TIMOB-26246] Android: Added Ti.UI.Window "extendSafeArea", "safeAreaPadding", and inset/notch support

jquick-axway opened the pull request:
**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-26246
- https://jira.appcelerator.org/browse/TIMOB-26427
- https://jira.appcelerator.org/browse/TIMOB-26459
- https://jira.appcelerator.org/browse/TIMOB-25810
- https://jira.appcelerator.org/browse/TIMOB-26460
- https://jira.appcelerator.org/browse/TIMOB-26442

**Summary:**
- [TIMOB-26246](https://jira.appcelerator.org/browse/TIMOB-26246) Added handling of Android P insets/notches/cutouts.
  * Added `Ti.UI.Window` property "extendSafeArea" to render content beneath insets.
  * Unlike iOS, this property is `false` by default on Android. This is because insets can be of any size and on side, making it more problematic.
- [TIMOB-26427](https://jira.appcelerator.org/browse/TIMOB-26427) Added new `Ti.UI.Window` property "safeAreaPadding".
  * Applies to Android 4.4 and newer OS versions when "extendSafeArea" is `true`.
  * Needed to detect size of insets for translucent StatusBar/NavBar and Android P notches.
  * Padding will be all zeros if "extendSafeArea" is `false`.
- [TIMOB-26459](https://jira.appcelerator.org/browse/TIMOB-26459) Added Android 4.4 constants for `Ti.UI.Window` property "windowFlags":
  * `Ti.UI.Android.FLAG_TRANSLUCENT_NAVIGATION`
  * `Ti.UI.Android.FLAG_TRANSLUCENT_STATUS`
- [TIMOB-25810](https://jira.appcelerator.org/browse/TIMOB-25810) Fixed bug where `ActionBar` height won't resize upon orientation changes.
- [TIMOB-26460](https://jira.appcelerator.org/browse/TIMOB-26460) Fixed bug where `Toolbar` height set to `Ti.UI.SIZE` won't resize upon orientation change.
- [TIMOB-26442](https://jira.appcelerator.org/browse/TIMOB-26442) Fixed bug where `Toolbar` with "extendBackground" to `true` is too tall if both status bar and nav bar are translucent.
- Modified DrawerLayout XML to not fit its "center" view under top insets:
  * The XML setting was wrong. It was intended to place left/right drawers under top inset, but it was only doing it main center view instead.
  * Better solution is to allow left/center/right layouts to overlap insets (like gmail app) and use a `Toolbar` with "extendBackground" `true` to position itself beneath top inset.

**Test Prerequisite:**
You must obtain an Android P device with a notch before doing the below tests.
If the device doesn't have a notch, then you can simulate it as follows:
1. Go to the Android P device's "Settings" screen.
2. Tap on "Systems" from the list.
3. Tap on "Advanced".
4. Tap on "Developer Options".
5. Tap on "Simulate a display with cutouts" to display a dialog.
6. Select "Double display cutout". (Will simulate notch on top and bottom.)

---
**Toolbar Test:**
1. Build and run test code attached to [TIMOB-26442](https://jira.appcelerator.org/browse/TIMOB-26442) on Android 4.4 or higher.
2. Verify that there is no extra padding below the `Toolbar`. (See [TIMOB-26442](https://jira.appcelerator.org/browse/TIMOB-26442) screenshots.)
3. Rotate the device to landscape.
4. Verify that the `Toolbar` height and font size shrinks. (This tests [TIMOB-26460](https://jira.appcelerator.org/browse/TIMOB-26460).)
5. Continue holding the device in landscape and back out of the app.
6. Relaunch the app in landscape.
7. Rotate the device to portrait.
8. Verify that the `Toolbar` height and font size increase in portrait.

---
**ActionBar Test:**
1. Create a default ClassicApp test project.
2. Hold an Android device in portrait form.
3. Build and run the app on that Android device.
4. Rotate the device to landscape.
5. Verify that the titlebar height and font size shrinks. (This tests [[TIMOB-25810](https://jira.appcelerator.org/browse/TIMOB-25810)].)
6. Continue holding the device in landscape and back out of the app.
7. Relaunch the app in landscape.
8. Rotate the device to portrait.
9. Verify that the titlebar height and font size increases in size.

---
**Translucent StatusBar and NavBar Test:**
1. Build and run the code attached to [TIMOB-26459](https://jira.appcelerator.org/browse/TIMOB-26459) on Android 4.4 or higher.
2. Verify that screen looks like what's shown in that ticket. App window with green border should fill the screen and top status bar and bottom nav bar should be overlaid on top.

_Note that Android 4.4 will show an opaque navigation bar when displayed landscape. This is a limitation on Google's end._

---
**Extend Safe-Area Window Test:**
1. Build and run `ExtendSafeAreaContainerTest.js` attached to [TIMOB-26427](https://jira.appcelerator.org/browse/TIMOB-26427).
2. Hold the device portrait.
3. Verify that a green rectangle with a blue border is between all insets/notches. (See: [image](https://jira.appcelerator.org/secure/attachment/65752/ExtendSafeAreaContainerPortrait.jpeg))
4. Rotate device to landscape where navigation bar is on the right side.
5. Verify that the green rectangle is still between the insets. (See: [image](https://jira.appcelerator.org/secure/attachment/65753/ExtendSafeAreaContainerLandscapeRight.jpeg))
6. Rotate device to the other landscape orientation without rotating to portrait. _(Do a barrel roll!)_
7.  Verify that the green rectangle is still between the insets. (See: [image](https://jira.appcelerator.org/secure/attachment/65754/ExtendSafeAreaContainerLandscapeLeft.jpeg))

---
**Extend Safe-Area ScrollView Test:**
1. Build and run `ExtendSafeAreaScrollViewTest.js` attached to [TIMOB-26427](https://jira.appcelerator.org/browse/TIMOB-26427).
2. Hold the device in portrait.
3. Scroll to the very top, if not already.
4. Verify that top-most view is not overlapped by the top status bar as seen [here](https://jira.appcelerator.org/secure/attachment/65749/ExtendSafeAreaScrollPortraitTop.jpeg).
5. Scroll to the very bottom.
6. Very bottom-most view is not overlapped by the bottom navigation bar as seen [here](https://jira.appcelerator.org/secure/attachment/65750/ExtendSafeAreaScrollPortraitBottom.jpeg).
7. Rotate to landscape where navigation bar is on the right side of the screen.
8. Scroll to the very top while landscape.
9. Verify that all rows are not overlapped by insets on the top/left/right sides as seen [here](https://jira.appcelerator.org/secure/attachment/65746/ExtendSafeAreaScrollLandscapeRightTop.jpeg).
10. Rotate device to the other landscape orientation without going to portrait. _(Do a barrel roll!)_
11. Scroll to the very top.
12. Verify that all rows are not overlapped by insets on the top/left/right sides as seen [here](https://jira.appcelerator.org/secure/attachment/65747/ExtendSafeAreaScrollLandscapeLeftTop.jpeg).
13. Scroll to the very bottom while landscape.
14. Verify that "Row 20" is shown without overlap at the bottom. Only top-most rows are overlapped by the top status bar as seen [here](https://jira.appcelerator.org/secure/attachment/65748/ExtendSafeAreaScrollLandscapeLeftBottom.jpeg).

---
**Extend Safe-Area TabGroup Test:**
1. Build and run `ExtendSafeAreaTabsTest.js` attached to [TIMOB-26427](https://jira.appcelerator.org/browse/TIMOB-26427).
2. Hold the device in portrait form and tap "Tab 1".
3. Verify that a green rectangle is displayed between the top tab bar and the bottom translucent navigation bar without overlap... and a red background fills the translucent areas as seen [here](https://jira.appcelerator.org/secure/attachment/65756/ExtendSafeAreaTab1Portrait.jpeg).
4. Tap "Tab 2" while portrait.
5. Verify Tab 2 display a green rectangle between the top tab bar and bottom navigation bar too... and an orange background fills the rest of the translucent area as shown [here](https://jira.appcelerator.org/secure/attachment/65757/ExtendSafeAreaTab2Portrait.jpeg).
6. Rotate device to landscape where navigation bar is on the right side.
7. Tap "Tab 1" while landscape.
8. Verify green rectangle is between all insets and the top tab bar as seen [here](https://jira.appcelerator.org/secure/attachment/65759/ExtendSafeAreaTab1LandscapeRight.jpeg).
9. Rotate device to the other landscape without going to portrait. _(Do a barrel roll!)_
10. Verify green rectangle is between all insets and the top tab bar as seen [here](https://jira.appcelerator.org/secure/attachment/65760/ExtendSafeAreaTab1LandscapeLeft.jpeg).
11. Tap "Tab 2" and verify green rectangle is between all insets and top tab bar.

---
**Extend Safe-Area DrawerLayout Test:**
1. Build and run `ExtendSafeAreaDrawerTest.js` attached to [TIMOB-26427](https://jira.appcelerator.org/browse/TIMOB-26427).
2. Hold the device portrait.
3. Verify that the top toolbar's title is below the status bar, but the toolbar's red background extends under the top status bar as seen [here](https://jira.appcelerator.org/secure/attachment/65739/ExtendSafeAreaDrawerMainPortrait.jpg).
4. Tap the hamburger button in the top-left corner to open the drawer.
5. Verify that the drawer's top blue toolbar's title is below the status bar, but the blue toolbar's background is extended under the status bar as seen [here](https://jira.appcelerator.org/secure/attachment/65741/ExtendSafeAreaDrawerPanelPortrait.jpg).
6. Rotate device to landscape.
7. Verify title is still below top status bar and shifted to the right if there is an Android P notch on the left side as seen [here](https://jira.appcelerator.org/secure/attachment/65742/ExtendSafeAreaDrawerPanelLandscape.jpg).
8. Press the back button to close the drawer.
9. Verify that main window's top red toolbar title is still below the status bar and shifted to the right if there is an Android P notch on the left side of the screen as seen [here](https://jira.appcelerator.org/secure/attachment/65740/ExtendSafeAreaDrawerMainLandscape.jpg).

jquick-axway committed:
[TIMOB-26246] Android: Added Ti.UI.Window "extendSafeArea", "safeAreaPadding", and inset/notch support

- [TIMOB-26246] Added handling of Android P inset/notch/cutout.
  * Added Ti.UI.Window property "extendSafeArea" to render content beneath insets.
- [TIMOB-26427] Added Ti.UI.Window property "safeAreaPadding".
  * Applies to Android 4.4 and newer OS versions when "extendSafeArea" is true.
  * Needed to detect size of insets for translucent StatusBar/NavBar and Android P notches.
- [TIMOB-26459] Added constants for Ti.UI.Window property "windowFlags".
  * Ti.UI.Android.FLAG_TRANSLUCENT_NAVIGATION
  * Ti.UI.Android.FLAG_TRANSLUCENT_STATUS
- [TIMOB-25810] Fixed bug where ActionBar height won't resize upon orientation changes.
- [TIMOB-26460] Fixed bug where Toolbar height set to Ti.UI.SIZE won't resize upon orientation change.
- [TIMOB-26442] Fixed bug where Toolbar with "extendBackground" to true is too tall if both status bar and nav bar are translucent.
- Modified DrawerLayout XML to not fit its "center" view under top insets.
  * The XML setting was wrong. It was intended to place left/right drawers under top inset, but it wasn't doing that.
  * Better solution is to allow left/center/right layouts to overlap insets and use a Toolbar with "extendBackground" true to position itself beneath inset.

jquick-axway committed:
Merge branch 'master' into TIMOB-26246

build commented:
<!--
  1 failure:  :disappointed_rel...
  0 warning: 
  1 messages
  
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.5.0 test /Users/build/jenkins/workspace/dk_titanium_mobile_PR-10383-YLLHVC3BZYNCRUJV3LCJHP2T7R5UZRWTK6DEYSQ6ABJTPGPLGKRA
> grunt

Running "appcJs:src:lintOnly" (appcJs) task

Running "eslint:src" (eslint) task

Running "checkFormat:ios" (checkFormat) task

Running "checkFormat:android" (checkFormat) task
Fatal error: Error: Formatting incorrect on "android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='18325' length='6'>&#10;								 </replacement>
</replacements>

Error: Formatting incorrect on "android/titanium/src/java/org/appcelerator/titanium/view/TiActivitySafeAreaMonitor.java", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='943' length='2'> </replacement>
<replacement offset='2931' length='11'> </replacement>
<replacement offset='2953' length='1'>&#10;									   </replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

jquick-axway committed:
Android: Updated code formatting for [TIMOB-26246]

ypbnv commented on android/titanium/src/java/org/appcelerator/titanium/proxy/TiWindowProxy.java:
```diff
@@ -39,6 +41,9 @@
 import android.view.Display;
 import android.view.View;
+import android.view.ViewParent;
+import android.view.Window;
```
Unused import.

ypbnv commented on android/titanium/src/java/org/appcelerator/titanium/view/TiActionBarStyleHandler.java:
```diff
@@ -0,0 +1,108 @@
+	}
+
+	/**
+	 * To be called by the owner when the activity's overriden onConfigurationChanged() method has been called.
```
Typo in "overridden".

ypbnv commented on android/titanium/src/java/org/appcelerator/titanium/view/TiActionBarStyleHandler.java:
```diff
@@ -0,0 +1,108 @@
+	/**
+	 * To be called by the owner when the activity's overriden onConfigurationChanged() method has been called.
+	 * Updates the ActionBar's height and font size base on the given configuration.
+	 * @param newConfig The udpated configuration applied to the activity.
```
Typo in "updated".

ypbnv commented on android/titanium/src/java/org/appcelerator/titanium/view/TiActionBarStyleHandler.java:
```diff
@@ -0,0 +1,108 @@
+				toolbarStyleHandler = new TiToolbarStyleHandler((Toolbar) view);
+			}
+		} catch (Exception ex) {
+		}
```
Maybe logging a warning or a debug level message here ? To make it easier for troubleshooting at some point if a problem comes from failing this try block. ( I am not a fan of empty catch blocks )

jquick-axway replied:
I agree. Done.

ypbnv commented on android/titanium/src/java/org/appcelerator/titanium/view/TiActivitySafeAreaMonitor.java:
```diff
@@ -0,0 +1,389 @@
+
+		// Calculate the safe-area, which is the region between the insets.
+		// Note: We must dynamically add ActionBar height here (if enabled) since its
+		//       visbility and height changes can only be detected via an onLayoutChange() event.
```
Typo in "visibility".

ypbnv commented on android/titanium/src/java/org/appcelerator/titanium/view/TiToolbarStyleHandler.java:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	/**
+	 * To be called by the owner when the activity/view's overriden onConfigurationChanged() method has been called.
```
Typo in "overridden".

ypbnv commented on android/titanium/src/java/org/appcelerator/titanium/view/TiToolbarStyleHandler.java:
```diff
@@ -0,0 +1,114 @@
+	/**
+	 * To be called by the owner when the activity/view's overriden onConfigurationChanged() method has been called.
+	 * Updates the toolbar's height and font size base on the given configuration.
+	 * @param newConfig The udpated configuration applied to the activity/view.
```
Typo in "updated".

ypbnv commented on android/titanium/src/java/org/appcelerator/titanium/view/TiToolbarStyleHandler.java:
```diff
@@ -0,0 +1,114 @@
+	 * Updates the toolbar's height and font size base on the given configuration.
+	 * @param newConfig The udpated configuration applied to the activity/view.
+	 */
+	public void onConfigurationChanged(Configuration newConfig)
```
The `newConfig` parameter seems to not be used in the method. It is used when the method is called from:
https://github.com/appcelerator/titanium_mobile/pull/10383/files#diff-9ea98253bfde40055a91a9df7a25cb33R52
and 
https://github.com/appcelerator/titanium_mobile/pull/10383/files#diff-9ce4241f5edf3cb3a0918ec183f9f59bR83
Maybe the former one may allow us to remove the parameter from the definition in `TiActionBarStyleHandler` too?

jquick-axway replied:
I made it this way to make it clear that this method should be called when the Activity/View object's `onConfigurationChanged()` method has been called. That's also partly why I decided to name the class a "handler" because it's like an event handler.

Also, I was originally thinking about optimizing this by only updating the ActionBar/Toolbar when the given configuration object indicated that the orientation changed. But `onConfigurationChanged()` is actually rarely called. So, I didn't think it was worth it.

Are you getting a compiler warning? I may need to add a `@SuppressWarnings("unused")`.

ypbnv replied:
I just saw it was unused. I am OK to leave it as is, matching the declaration makes sense for it being more clear what is happening.

ypbnv reviewed (approved):
Awesome PR! Besides a few minor things I have commented on it looks good!

jquick-axway committed:
Android: Updated [TIMOB-26246] based on feedback

jquick-axway committed:
Android: Modified [TIMOB-26246] to support multiple toolbars which extend their background

- Fixed issue where if multiple toolbars had "extendBackground" true, it would only apply to 1st one.
  * Had to override default behavior to prevent insets from being consumed.
- Fixed regression where toolbar "extendBackground" true would no longer work unless window "extendSafeArea" was also true.
  * Caused by root Titanium view calling setFitsSystemWindows(true) on startup, instead of default false setting.

longton95 committed:
Merge branch 'master' into TIMOB-26246

ewanharris commented on apidoc/Titanium/UI/Window.yml:
```diff
@@ -441,24 +441,25 @@ properties:
     type: Boolean
-    default: true
-    since: 6.3.0
+    default: `true` on iOS. `false` on Android.
```
This is causing the docgen validation to error out as it's a string (which fails the parsing of Window.yml, throwing all the other parse errors)
It's super easy to allow the validation script to allow us to set `{android: false, ios: true}` here, but I have no idea how this then plays into the generation of the various templates (html, jdsuck, typescript etc.). @jawa9000 do you know how I can check the output?

jquick-axway replied:
Thanks. I'll give it a go.

jquick-axway committed:
Fixed API doc parse error caused by [TIMOB-26246]

jquick-axway committed:
Merge branch 'master' into TIMOB-26246

longton95 committed:
Merge branch 'master' into TIMOB-26246

longton95 reviewed (approved):
**FR Passed**

Verified using the test cases above,  `extendSafeArea` and `safeAreaPadding` work as expected with the notch supports


**ENV**

```
Pixel 2 XL (9.0)
Emulator (9.0, 6.0, 4.4.2)
Appc NPM: 4.2.13
Appc CLI: 7.0.7-9
Ti CLI: 5.1.1
Node: 8.12.0
NPM: 6.4.1
```

build commented:
<!--
  1 failure:  Tests have failed...
  0 warning: 
  1 messages
  1 markdown notices
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Tests have failed, see below for more information.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-10383/18/artifact/dist/mobilesdk-7.5.0.v20181025074022-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>

### Tests: 

| Classname | Name | Time | Error |
| --- | --- | --- | --- |
| android.Titanium.UI.TextField | focus-blur-bubbles | 6.127 | timeout of 5000ms exceeded |
| ios.Titanium.Media.AudioPlayer | .duration | 60.007 | expected -2147483648 to be within 45250..45500 |

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>
